main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -980,18 +988,104 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
+
+      classOpt match {
+        case Some(clazz) =>
+          // Case 1.2: Found in the current namespace.
+          Some(clazz)
+        case None =>
+          // Case 1.1: Try the global namespace.
+          root.classes.getOrElse(Name.RootNS, Map.empty).get(qname.ident.name)
+      }
+    } else {
+      // Case 2: Qualified. Lookup in the given namespace.
+      root.classes.getOrElse(qname.namespace, Map.empty).get(qname.ident.name)
+    }
+  }
+  /**
+    * Finds the def with the qualified name `qname` in the namespace `ns0`.
     */
   def lookupDef(qname: Name.QName, ns0: Name.NName, root: NamedAst.Root): Validation[NamedAst.Def, ResolutionError] = {
     val defOpt = tryLookupDef(qname, ns0, root)
 
     defOpt match {
       case None => ResolutionError.UndefinedName(qname, ns0, qname.loc).toFailure
-      case Some(d) => getDefIfAccessible(d, ns0, qname.loc)
+      case Some(defn) =>
+        if (isDefAccessible(defn, ns0)) {
+          defn.toSuccess
+        } else {
+          ResolutionError.InaccessibleDef(defn.sym, ns0, qname.loc).toFailure
+        }
+    }
+  }
+
+  /**
+    * Finds the def or sig with the qualified name `qname` in the namespace `ns0`.
+    */
+  def lookupDefSig(qname: Name.QName, ns0: Name.NName, root: NamedAst.Root, sigs: Map[Name.NName, Map[String, NamedAst.Sig]]): Validation[Either[NamedAst.Def, NamedAst.Sig], ResolutionError] = {
+    val defOpt = tryLookupDef(qname, ns0, root)
+    val sigOpt = tryLookupSig(qname, ns0, sigs)
+
+    (defOpt, sigOpt) match {
+      // Case 1: Can't find the name anywhere
+      case (None, None) => ResolutionError.UndefinedName(qname, ns0, qname.loc).toFailure
+      // Case 2: Can find only a def with the name
+      case (Some(defn), None) =>
+        if (isDefAccessible(defn, ns0)) {
+          Left(defn).toSuccess
+        } else {
+          ResolutionError.InaccessibleDef(defn.sym, ns0, qname.loc).toFailure
+        }
+      // Case 3: Can find only a sig with the name
+      case (None, Some(sig)) =>
+        if (isSigAccessible(sig, ns0)) {
+          Right(sig).toSuccess
+        } else {
+          ResolutionError.InaccessibleSig(sig.sym, ns0, qname.loc).toFailure
+        }
+      // Case 4: Can find both with the name
+      case (Some(defn), Some(sig)) =>
+        (isDefAccessible(defn, ns0), isSigAccessible(sig, ns0)) match {
+          // Case 4.1: Neither is accessible
+          case (false, false) => ResolutionError.InaccessibleDef(defn.sym, ns0, qname.loc).toFailure // MATT need something for if both are inaccessible?
+          // Case 4.2: Only the def is accessible
+          case (true, false) => Left(defn).toSuccess
+          // Case 4.3: Only the sig is accessible
+          case (false, true) => Right(sig).toSuccess
+          // Case 4.4: Both are accessible
+          case (true, true) => ResolutionError.AmbiguousName(qname, ns0, List(defn.loc, sig.loc), qname.loc).toFailure
```

Comment:
We should discuss this. I see two options:

(1) Resolution happens before visibility checking.
(2) Resolution takes visibility checking into account.

I think (1) is cleanest, but (2) is of course more flexible.

With (1) you also have the advantage that if there is an error then you definitely know that its because of a visibility issue. With (2) you risk that the program "works" by selecting the wrong function (!?).

This specific issue should, at least in principle, be discussed and decided before we merge I think :)

Reply:
I agree, (1) seems like the safer option.